### PR TITLE
Add fastpath test for mask check flag

### PR DESCRIPTION
**Pull request by yoavnavon**
Summary: Check that fastpath is taken, which type (sparsity fastpath or normal) for mask that is aligned and one that is not.

Test Plan: buck test caffe2/test:test_transformers

Differential Revision: D38259928

### CODE REVIEW

File: test/test_transformers.py [hunk 1]
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
...
 
+    @unittest.skipIf(TEST_WITH_CROSSREF, 'Fastpath not available with crossref')
+    def test_mask_check_fastpath(self):
+        """
+        Test that fastpath is executed independently of the mask that is passed.
+        If the passed mask is left aligned, test that nested tensors are used (sparsity fastpath),
+        otherwise use fastpath with traditional tensors.
+        """
+
+        with torch.no_grad():
```
Comment (jbschlosser) [inline test/test_transformers.py:742]: style nit: you could alternatively decorate the test with `@torch.no_grad()`

File: test/test_transformers.py [hunk 2]
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
...
+            x = torch.Tensor([[[1, 2], [3, 4], [5, 6]]]).to(torch.float)
+
+            def _test_fastpath(model, mask, mock_return_value, nested_tensors=True):
+                with patch('torch._transformer_encoder_layer_fwd') as fastpath_mock:
+                    fastpath_mock.return_value = mock_return_value
+                    model(x, src_key_padding_mask=mask)
+                    self.assertTrue(fastpath_mock.called)
```
Comment (jbschlosser) [inline test/test_transformers.py:749]: don't we expect this to be False for certain inputs?

File: test/test_transformers.py [hunk 3]
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
...
+        Test that fastpath is executed independently of the mask that is passed.
+        If the passed mask is left aligned, test that nested tensors are used (sparsity fastpath),
+        otherwise use fastpath with traditional tensors.
+        """
+
+        with torch.no_grad():
+            x = torch.Tensor([[[1, 2], [3, 4], [5, 6]]]).to(torch.float)
+
+            def _test_fastpath(model, mask, mock_return_value, nested_tensors=True):
+                with patch('torch._transformer_encoder_layer_fwd') as fastpath_mock:
```
Comment (jbschlosser) [inline test/test_transformers.py:746]: `_transformer_encoder_layer_fwd` represents the "sparsity fastpath", correct? Maybe we should name the mock with `sparsity` to avoid ambiguity as to which fastpath is being mocked out

File: test/test_transformers.py [hunk 4]
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
...
+
+            model = torch.nn.TransformerEncoder(encoder_layer, num_layers=2, enable_nested_tensor=True, mask_check=False)
+            model.eval()
+
+            # Mask check disabled with aligned mask results in sparsity fastpath
+            _test_fastpath(model, aligned_mask, nested_tensor_return_value, nested_tensors=True)
+
+            # TODO: Check disabled with non aligned mask should not raise any error
+            # but return unmeaningful values.
+            # _test_fastpath(model, not_aligned_mask, nested_tensor_return_value, nested_tensors=True)
```
Comment (jbschlosser) [inline test/test_transformers.py:785]: what are we waiting on to address this TODO? Do we need to avoid a second alignment check that happens in the sparsity fastpath?

**Review by jbschlosser — COMMENTED** [review]
Thanks for the tests! Just a few comments below to make sure we're validating the correct behavior

File: test/test_transformers.py (hunk 3 above)
Comment (yoavnavon) [reply test/test_transformers.py:746]: Not exactly, `_transformer_encoder_layer_fwd` represents both the "sparsity fastpath" and normal fastpath. When the method is called with nested tensors, that is the sparsity fastpath, otherwise it is normal fastpath.

Comment (jbschlosser) [reply test/test_transformers.py:746]: Okay gotcha. How are we validating whether the "sparsity fastpath" or "normal fastpath" is taken? Is this based on the args passed to `_transformer_encoder_layer_fwd`?

File: test/test_transformers.py (hunk 2 above)
Comment (yoavnavon) [reply test/test_transformers.py:749]: Yes, but not in the context of the `mask_check` flag which is what we are testing right now. There are multiple cases where fastpath is not taken (like using grad, or model in training mode) but right now we are assuming those conditions are met to test just that flag.

File: test/test_transformers.py (hunk 4 above)
Comment (yoavnavon) [reply test/test_transformers.py:785]: @mikekgfb  What do you think about this?

File: test/test_transformers.py (hunk 1 above)
Comment (yoavnavon) [reply test/test_transformers.py:742]: Nice!

File: test/test_transformers.py (hunk 2 above)
Comment (jbschlosser) [reply test/test_transformers.py:749]: Cool, I'm good with this

File: test/test_transformers.py [hunk 5]
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
...
+        with torch.no_grad():
+            x = torch.Tensor([[[1, 2], [3, 4], [5, 6]]]).to(torch.float)
+
+            def _test_fastpath(model, mask, mock_return_value, nested_tensors=True):
+                with patch('torch._transformer_encoder_layer_fwd') as fastpath_mock:
+                    fastpath_mock.return_value = mock_return_value
+                    model(x, src_key_padding_mask=mask)
+                    self.assertTrue(fastpath_mock.called)
+                    for call_args, _ in fastpath_mock.call_args_list:
+                        self.assertEqual(call_args[0].is_nested, nested_tensors)
```
Comment (jbschlosser) [inline test/test_transformers.py:751]: Suggest adding a short comment for clarity here mentioning that receiving NT args corresponds with the "sparsity fastpath"

File: test/test_transformers.py [hunk 6]
```diff
@@ -724,6 +731,58 @@ def rand_tensor(*shape):
             if dropout_p == 0.0 or device == 'cpu':
                 self.assertEqual(actual, expected)
 
+    @unittest.skipIf(TEST_WITH_CROSSREF, 'Fastpath not available with crossref')
+    def test_mask_check_fastpath(self):
+        """
+        Test that fastpath is executed independently of the mask that is passed.
+        If the passed mask is left aligned, test that nested tensors are used (sparsity fastpath),
```
Comment (jbschlosser) [inline test/test_transformers.py:738]:
```suggestion
        If the passed mask is left aligned or mask_check=False, test that nested tensors are used (sparsity fastpath),
```

**Review by jbschlosser — APPROVED** [review]
LGTM
